fsa_server.R

```diff
@@ -0,0 +1,120 @@
+bracket_regex <- "\\[|\\]|\\<|\\>|\\{|\\}"
+
+fread_type_mapping <- list(
+  "c" = "character",
+  "d" = "double",
+  "i" = "integer",
+  "T" = "POSIXct",
+  "l" = "logical",
+  "D" = "Date",
+  "n" = "numeric"  # Skip column
+)
+
+# Function to detect non-UTF-8 characters and brackets
+detect_problematic_and_bracket_characters <- function(dt, file) {
+  problematic_characters <- list()
+  for (col in names(dt)) {
+    if (is.character(dt[[col]])) {
+      # Detect non-UTF-8 characters
+      invalid_utf8_rows <- which(!stringi::stri_enc_isutf8(dt[[col]]))
+      if (length(invalid_utf8_rows) > 0) {
+        problematic_characters$utf8_issues <- lapply(invalid_utf8_rows, function(row) {
+          list(
+            row = row, 
+            col = col,
+            chars= iconv(
```

Comment:
This converts windows-encoded characters, like smart-quotes, to the UTF-8 approximation so that the user will see the actual character in the export. Otherwise, they'd see the UTF-8 interpreted.

NOTE 12/13/24: AS and GD decided better to silently convert and not flag because the actual characters may be legitimate UTF-8 characters (like smart quotes are), but the FILE may be encoded as Windows. Helping the user encode a file as UTF-8 will be difficult and it's not their fault in the first place since they just want the characters that ARE VALID UTF-8 anyway